intellij-plugin-verifier/verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/resolution/Methods.kt

```diff
@@ -0,0 +1,100 @@
+package com.jetbrains.pluginverifier.verifiers.resolution
+
+import com.jetbrains.plugin.structure.classes.resolvers.ResolutionResult
+import com.jetbrains.plugin.structure.classes.resolvers.Resolver
+import org.objectweb.asm.Opcodes
+import org.objectweb.asm.tree.AbstractInsnNode
+import org.objectweb.asm.tree.ClassNode
+import org.objectweb.asm.tree.MethodInsnNode
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+private val LOG: Logger = LoggerFactory.getLogger(Method::class.java)
+
+/**
+ * Indicates a binary class name where packages are delimited by '/' instead of '.'.
+ * This is in line with Java Language Specification.
+ */
+typealias BinaryClassName = String
+
+fun Method.isOverriding(anotherMethod: Method): Boolean =
+  nonStaticAndNonFinal(anotherMethod)
+    && sameName(this, anotherMethod)
+    && sameParametersAndReturnType(this, anotherMethod)
+    && sameOrBroaderVisibility(this, anotherMethod)
+
+data class MethodInClass(val method: Method, val klass: ClassFile)
+
+private fun nonStaticAndNonFinal(anotherMethod: Method) = !anotherMethod.isStatic && !anotherMethod.isFinal
+
+private fun sameName(method: Method, anotherMethod: Method) = method.name == anotherMethod.name
+
+private fun sameParametersAndReturnType(method: Method, anotherMethod: Method) =
+  method.descriptor == anotherMethod.descriptor
+
+private fun sameOrBroaderVisibility(method: Method, anotherMethod: Method): Boolean {
+  return method.visibilityRating >= anotherMethod.visibilityRating
+}
+
+private val Method.visibilityRating: Int
+  get() = when {
+    isPublic -> 3
+    isProtected -> 2
+    isPackagePrivate -> 1
+    isPrivate -> 0
+    else -> -1
+  }
+
+/**
+ * Search for all methods in the parent hierarchy that the receiver overrides.
+ * @return list of methods, sorted from bottom-to-top in the inheritance hierarchy.
+ */
+fun Method.searchParentOverrides(resolver: Resolver): List<MethodInClass> {
+  return mutableListOf<MethodInClass>().apply {
+    searchParentOverrides(resolver) { klass: ClassFile, method: Method ->
+      add(MethodInClass(method, klass))
+    }
+  }
+}
+
+fun Method.searchParentOverrides(resolver: Resolver, matchHandler: (ClassFile, Method) -> Unit) {
+  var superClassFqn = containingClassFile.superName ?: return
```

Comment:
does it skip `java.lang.Object` automatically?

Reply:
By default, `java.lang.Object` is not skipped. Added a parameter to enable skipping, disabled by default.

Reply:
why would we ever want to check java.lang.Object? it is not annotated with OverrideOnly and never will be

Reply:
I don't see a point to keep this parameter

Reply:
Parameter removed. Documentation added to indicate this behaviour.